### PR TITLE
Add option to preserve/resolve symlinks

### PR DESCRIPTION
**Pull request by ajafff**
## Description
Add option `preserveSymlinks` everywhere:
* resolve/index.js
  * works for both AMD and CJS
* extract.js
* validate.js
* defaults.json
* jsonschema (twice, don't know why)
* CLI
* docs
* tests

The option (currently) defaults to `true` which leaves the old behavior unchanged. This can and probably should be changed in a future major version to be consistent with Node's default behavior.
To disable this option, either specify it in the config file or use the CLI argument `--preserve-symlinks false`.

## Motivation and Context
Fixes: #33 

## How Has This Been Tested?
Tested in the unit tests. (needed to add a devDependency on `symlink-dir` for easier handling of symlinks on windows)
Also tested locally with in my project https://github.com/fimbullinter/wotan. Tested config file and CLI option.

## Types of changes

- ~~[ ] Bug fix (non-breaking change which fixes an issue)~~
- [x] New feature (non-breaking change which adds functionality)
- ~~[ ] Breaking change (fix or feature that would cause existing functionality to change)~~

## Checklist:

- [x] The code I add will be subject to [The MIT license](../LICENSE), and I'm OK with that.
- [x] The code I've added is my own original work.
- [x] My code follows the code style of this project. (I hope so, this is not always obvious)
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](./CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by sverweij** [comment]
Thanks! It looks very complete at a first glance. I'll try to review it more closely tomorrow.

**Comment by sverweij** [comment]
I'll do the review in chunks so it is easier to follow & comment on

#### Your question about the jsonschema:
> jsonschema (twice, don't know why)
Good point;
- The _rule set_ has it in the options section for obvious reasons; 
- Less obvious is why it's in the (json) _output file_. That file has a section with the command line options used to generate it. I put it there for support, and the ability to put it in custom reporters. 
Maybe some day I should put the shared definitions in a separate schema and link to it from the other two schemas to be more DRY.

**Comment by sverweij** [comment]
### command line & config option
#### switch vs option with a parameter
I see you didn't go for the 'switch', parameter-less of command line option, and I understand the dilemma:
- putting in a switch with the proper name (`--preserve-symlinks`, to mirror the one in node) would introduce a breaking change because leaving it out imply would imply the default for it would be `false`
- putting in a switch that doesn't introduce a breaking change (`--follow-symlinks`) could be confusing both in code and to people using the command line.

... so you did the only sensible thing left and opted for the proper name + config parameter in order to not introduce a breaking change. I appreciate that!

I've given it some thought and I'd rather do the breaking change right now (and 'semver-major' the release number) - which is consistent with your advice to stay aligned with node's default behaviour. 

#### Small things
I've taken the liberty to update the PR with some nitpicky things as well:
- the (line) test coverage fell below 100% because there was no test covering the command line validity check - this is partly my fault for not configuring a minimum coverage in the ci (fixed).
- The first and last options that appear when running depcruisie without parameters (/ or with --help) are the easiest to find. Which is why the `--init` option was on the last option (fixed).
- I've also added a thing so the followSymlinks appear in the `optionsUsed` in the extraction output object (& the json).

File: src/extract/extract.js [hunk 1]
```diff
@@ -41,7 +42,8 @@ module.exports = (pFileName, pOptions) => {
             {
                 baseDir: process.cwd(),
                 moduleSystems: ["cjs", "es6", "amd"],
-                tsPreCompilationDeps: false
+                tsPreCompilationDeps: false,
+                preserveSymlinks: true
```
Comment (ajafff) [inline src/extract/extract.js]: this should be `false` to match the docs?

File: test/extract/fixtures/amd.json [hunk 2]
```diff
@@ -247,5 +247,67 @@
...
+    },
+    {
+        "title": "require a symlinked module",
+        "input": {
+            "fileName": "test/extract/fixtures/amd/symlink.js"
+        },
+        "expected": [
+            {
+                "module": "../symlinked",
+                "resolved": "test/extract/fixtures/symlinked/index.js",
```
Comment (ajafff) [inline test/extract/fixtures/amd.json]: this will change when the default is set to `false`

File: test/extract/fixtures/cjs.json [hunk 3]
```diff
@@ -259,5 +259,89 @@
...
+    },
+    {
+        "title": "require a symlinked module",
+        "input": {
+            "fileName": "test/extract/fixtures/cjs/symlink.js"
+        },
+        "expected": [
+            {
+                "module": "../symlinked",
+                "resolved": "test/extract/fixtures/symlinked/index.js",
```
Comment (ajafff) [inline test/extract/fixtures/cjs.json]: same as above

**Pull request merged**